datafusion/expr/src/expr_schema.rs

```diff
@@ -453,6 +455,26 @@ impl ExprSchemable for Expr {
                 }
                 _ => Ok(Expr::Cast(Cast::new(Box::new(self), cast_to_type.clone()))),
             }
+        } else if matches!(
+            (&this_type, cast_to_type),
+            (DataType::Int32 | DataType::Int64, DataType::Interval(_))
+        ) {
+            // Convert integer (days) to the corresponding DayTime interval
+            match self {
+                Expr::Literal(ScalarValue::Int32(Some(days))) => {
+                    Ok(Expr::Literal(ScalarValue::IntervalDayTime(Some(
+                        arrow_buffer::IntervalDayTime::new(days, 0),
+                    ))))
+                }
+                Expr::Literal(ScalarValue::Int64(Some(days))) => {
+                    Ok(Expr::Literal(ScalarValue::IntervalDayTime(Some(
+                        arrow_buffer::IntervalDayTime::new(days as i32, 0),
+                    ))))
+                }
+                _ => plan_err!(
+                    "Cannot automatically convert {this_type:?} to {cast_to_type:?}"
+                ),
+            }
```

Comment:
This works for literals only and is immediate.
For other types all the function does is create cast expression. Why would we want to special-case here?

Reply:
What's the alternative? You can't cast an integer to an interval in arrow. 

You are right; it only works for literals. Do you have thoughts on how to make it work for non also?

Reply:
if a Cast is supposed to work, it should work regardless how it's weaved into the plan.
`Expr.cast_to` should not have any additional logic (other than checks) vs creating `Expr::Cast` directly.

However, we should first decide whether integer should be castable to interval at all.

It's not castable in PostgreSQL
```
postgres=# SELECT 10::interval;
ERROR:  cannot cast type integer to interval
LINE 1: SELECT 10::interval;
                 ^
```

so I am inclined it should not be castable in DF either.


Which means, the `date + int` cannot be converted (desugared) to `date + interval` using ordinary cast expression (as such shouldn't exist), but it can be converted (desugared) by some other means (eg with a custom conversion function or combination of expressions).

Alternatively, we defer `date + int`  until the execution, eg somewhere in https://github.com/apache/datafusion/blob/cc43766104cf66b8dbf431cbf6bdf25eda6ba274/datafusion/physical-expr/src/expressions/binary.rs#L301
but i'd prefer to desugar earlier and avoid separate code path in the `BinaryExpr` physical expr

Reply:
Perhaps we require a new scalar function, `make_interval`, which can convert integers into intervals.
Then, similar to [`make_array`](https://github.com/apache/datafusion/blob/bfabd48de2e0deccbd8ce595936ad24f18baca65/datafusion/functions-nested/src/planner.rs#L101),  perform transformations within the SQL planner. 
Transform `4 + to_date('1970-01-03')` to `make_interval(4) + to_date('1970-01-03')` inside `plan_binary_op`

Reply:
Yes @jonahgao something like that is the approach! I've pinged the make_interval issue:

https://github.com/apache/datafusion/issues/6951